File: ABC086C.ml
(* ABC086C - Traveling *)

type coord = { x : int; y : int }

let distance a b = abs (a.x - b.x) + abs (a.y - b.y)

let solver (txy : (int * coord) array) =
  let zero = (0, { x = 0; y = 0 }) in
  let walk previous current =
    let dt = fst current - fst previous in
    let dist = distance (snd current) (snd previous) in
    if dt < dist then raise Not_found
    else if (dt - dist) mod 2 = 0 then current
    else raise Not_found
  in
  try Array.fold_left walk zero txy |> Fun.const "Yes" with Not_found -> "No"

let () =
  let n = read_int () in
  let init _ =
    Scanf.sscanf (read_line ()) "%d %d %d" (fun t x y -> (t, { x; y }))
  in
  Array.init n init |> solver |> print_endline

File: ABC086C.cpp
// ABC086C - Traveling
#include <iostream>
#include <tuple>
#include <vector>

using XY = std::pair<long, long>;

long distance(XY a, XY b)
{
    const auto [ax, ay] = a;
    const auto [bx, by] = b;
    return std::abs(ax - bx) + std::abs(ay - by);
}

using TXY = std::pair<long, XY>;

bool solver(std::vector<TXY> &A)
{
    TXY previous = {};
    for (const auto &[t, xy] : A)
    {
        auto dt = t - std::get<0>(previous);
        auto dist = distance(xy, std::get<1>(previous));
        if (dt < dist)
            return false;
        else if ((dt - dist) % 2 != 0)
            return false;
        previous = {t, xy};
    }
    return true;
}

int main()
{
    long N;
    std::cin >> N;
    std::vector<TXY> A(N);
    for (auto &a : A)
    {
        long n, x, y;
        std::cin >> n >> x >> y;
        a = {n, {x, y}};
    }
    std::cout << (solver(A) ? "Yes" : "No") << std::endl;
    return 0;
}
